core/src/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,685 @@
+                    .message
+                    .static_account_keys()
+                    .get(0)
+                    .ok_or(DeserializedPacketError::VoteTransactionError)?;
+                let slot = vote_state_update.last_voted_slot().unwrap_or(0);
+
+                Ok(Self {
+                    vote: Some(vote),
+                    slot,
+                    pubkey,
+                    vote_source,
+                    forwarded: false,
+                })
+            }
+            _ => Err(DeserializedPacketError::VoteTransactionError),
+        }
+    }
+
+    pub fn get_vote_packet(&self) -> Rc<ImmutableDeserializedPacket> {
+        self.vote.as_ref().unwrap().clone()
+    }
+
+    pub fn pubkey(&self) -> Pubkey {
+        self.pubkey
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+
+    pub fn is_forwarded(&self) -> bool {
+        // By definition all gossip votes have been forwarded
+        self.forwarded || matches!(self.vote_source, VoteSource::Gossip)
+    }
+
+    pub fn is_vote_taken(&self) -> bool {
+        self.vote.is_none()
+    }
+
+    pub fn take_vote(&mut self) -> Option<Rc<ImmutableDeserializedPacket>> {
+        self.vote.take()
+    }
+}
+
+pub fn deserialize_packets<'a>(
+    packet_batch: &'a PacketBatch,
+    packet_indexes: &'a [usize],
+    vote_source: VoteSource,
+) -> impl Iterator<Item = LatestValidatorVotePacket> + 'a {
+    packet_indexes.iter().filter_map(move |packet_index| {
+        LatestValidatorVotePacket::new(packet_batch[*packet_index].clone(), vote_source).ok()
+    })
+}
+
+// TODO: replace this with rand::seq::index::sample_weighted once we can update rand to 0.8+
+// This requires updating dependencies of ed25519-dalek as rand_core is not compatible cross
+// version https://github.com/dalek-cryptography/ed25519-dalek/pull/214
+pub(crate) fn weighted_random_order_by_stake<'a>(
+    bank: &Arc<Bank>,
+    pubkeys: impl Iterator<Item = &'a Pubkey>,
+) -> impl Iterator<Item = Pubkey> {
+    // Efraimidis and Spirakis algo for weighted random sample without replacement
+    let staked_nodes = bank.staked_nodes();
+    let mut pubkey_with_weight: Vec<(f64, Pubkey)> = pubkeys
+        .filter_map(|&pubkey| {
+            let stake = staked_nodes.get(&pubkey).copied().unwrap_or(0);
+            if stake == 0 {
+                None // Ignore votes from unstaked validators
+            } else {
+                Some((thread_rng().gen::<f64>().powf(1.0 / (stake as f64)), pubkey))
+            }
+        })
+        .collect::<Vec<_>>();
+    pubkey_with_weight.sort_by(|(w1, _), (w2, _)| w1.partial_cmp(w2).unwrap());
+    pubkey_with_weight.into_iter().map(|(_, pubkey)| pubkey)
+}
+
+pub struct VoteBatchInsertionMetrics {
+    num_dropped_gossip: usize,
+    num_dropped_tpu: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct LatestUnprocessedVotes {
+    latest_votes_per_pubkey: RwLock<HashMap<Pubkey, Arc<RwLock<LatestValidatorVotePacket>>>>,
```

Comment:
Is there a size limit on the number of pubkeys this can hold before we start dropping?

Reply:
Not currently, but I can add a capacity - which pubkeys would we decide to drop though?

Reply:
Will address it as part of #27409

Reply:
kk just gotta make sure if we add removal of pubkeys this doesn't become an issue: https://github.com/solana-labs/solana/pull/27558#discussion_r970126258